python/tests/expressions/test_expressions_base.py

```diff
@@ -841,3 +924,290 @@ def test_not_expression_binding(unbound_not_expression, expected_bound_expressio
     """Test that visiting an unbound NOT expression with a bind-visitor returns the expected bound expression"""
     bound_expression = base.visit(unbound_not_expression, visitor=base.BindVisitor(schema=table_schema_simple))
     assert bound_expression == expected_bound_expression
+
+
+def test_bound_boolean_expression_visitor_ignore_is_nan_for_string_type():
+    """Test ignoring IsNan expressions when the field type is a String"""
```

Comment:
This is suspicious. You can't bind `IsNaN` to a string, so this should never happen. In addition, I don't see why the visitor would handle this any differently. There may be a runtime failure because the expression is improperly bound (users should never do this) but the visitor shouldn't simply ignore it.

Reply:
I see, I noticed the behavior so included this test but what you're saying makes sense. So is it safe to just remove this test and expect that the higher-level APIs would prevent an `IsNaN` expression on a string type? Or do you feel that we need to add some validation here? (sounds like we don't need that validation but just want to confirm)
```py
@visit_bound_predicate.register(BoundIsNaN)
def _(expr: BoundIsNaN, visitor: BoundBooleanExpressionVisitor[T]) -> T:
    # validate that `expr.term.field.field_type` is an appropriate type
    return visitor.visit_is_nan(term=expr.term)
```

Reply:
We don't need validation in the visitor. We should assume that the bound expression has been created correctly by the binding process and not by people instantiating the expressions directly. Maybe we should also move these into `bound.py` to avoid exposing them?

We also need to understand why the expression was skipped. That sounds like a bug to me. I don't think there's anything that checks.

Reply:
@samredai, we still need to find out what's going on here. This appears to be ignoring parts of the bound expression.

Reply:
Okay, I looked at what's happening here and it is a bit misleading because we are rewriting expressions in the expression constructors. `BoundIsNaN(...)` actually returns `AlwaysFalse()` because it cannot be true for a string. Then `Not(AlwaysFalse())` returns `AlwaysTrue()` and `And` returns its left argument because the right argument is a constant.

In the end, this is testing expression construction that simplifies automatically, not the visitor. I'm going to just remove this test since it looks like everything functions properly.